### PR TITLE
Support for bounding box text edges

### LINKED ISSUE #112: Math limit formatting overlap
## Code

```
sup_(x in P_i)
```

## LaTeX
Screenshot 2023-03-22 at 7 52 03 AM

## Typst

Screenshot 2023-03-21 at 11 00 37 PM

### PR DESCRIPTION
**Pull request by damaxwell**
Closes #112.

Text in Typst has a notion of upper and lower edges, which can be specified by global font metrics (x-height, descender, etc.) In a math environment, where stuff is being laid out in a genuinely 2-d fashion, you want the upper and lower edges to be determined by the tight bounding box of the text instead so that there isn't excess space above all lower-cap words, or collisions due to descenders.

This PR addresses this by introducing two new values for a text edge, `bbox-top` and  `box-bottom`, which do just what they suggest.  In math layout, text is set to use these values to get the right size.

There was one additional change needed. Math layout used to use `Content::layout` for text, which ends up creating a `FlowElem` at the top level, which forgets about the baseline of its contents.  But the baseline is critical for math layout, so we render a `ParElem` explicitly instead of generic `Content` when laying out a `TextElem`.

**Before**
image

**After**
image

## Issues
1. No caching is done for glyph bounding boxes, so there is time inefficiency.  This could be added if desired, but for now it seemed like premature optimization..

2. The parameters for `text(top-edge:"...")` are a bit clumsy.  Already you can say things like 
 ```
#set text(top-edge:"baseline", bottom-edge:"x-height")
``` 
and it's not clear to me what this is actually supposed to do. The new options compound the awkwardness.  Now we have `#set text(top-edge:"bbox-top")`, which is a little verbose and you can say weird things like `#set text(top-edge:"bbox-bottom",bottom-edge:"bbox-top:")`. I'm not sure what this is supposed to do, regardless of what it actually does.

3.  A `TextElem` is always rendered inside of a paragraph. What is the width of a paragraph containing a `TextElem` inside of math layout? In the previous implementation, it was the width of the region the equation element is being laid out into.  But this is a little arbitrary, and if line breaks are actually encountered, the paragraph likely overfills the bounding math element anyway since it's probably not lined up with the container's left edge.  In the current implementation, `TextElem` inside of math simply never line breaks: it's laid out in to a region infinitely wide. This is essentially what `TeX` does with its `\hbox`s.

4. I wasn't sure how to set the text edge style for a `TextElem` in math mode programmatically and yet still allow this to be overridden by an insistent user. I used `Content::styled` but this seems to wipe out the text edge settings even for text elements created with explicit top/bottom edges.

## Tests
A number of tests images needed to be updated because text elements changed size.  In all cases, the resulting spacing was closer to that of `TeX`, or was negligibly different from the prior image due to very small differences between `bbox-top` versus `cap-height`, or was more consistent.  This last category includes things like the following behavior in the prior version: "if a `y` appears as a math character, it's descender affects the size of the `EquationElem` but if the same `y` appears as part of a larger `TextElem`, the descender is not accounted for"

### CODE REVIEW

**Comment by Enivex** [comment]
In that case there needs to be some mechanism to align limits vertically for consecutive operators (like \adjustlimits from mathtools in latex), because your example does not look great.

**Comment by damaxwell** [comment]
Well, beauty is in the eye of the beholder, isn't it.  Here's LaTeX, out of the box:
image
The spacing and layout looks **remarkably close** to what Typst produced.  Which is darn cool.

TeX (and LaTeX sitting on top of it) uses bounding boxes for determining the sizes of objects.  Every glyph to TeX when doing layout is just its bounding box (page 65 of the TeXBook has a nice visual of this) along with some other font metrics.  Appendix G of that book describes the math layout algorithm.  The spacing parameters on page 447 in that appendix have been encoded in a modern OpenType math font in its `math` table.  Those parameters are predicated on the text being described by tight bounding boxes.  Typst had done a good job with spacing in many scenarios already because texts runs of single glyphs were treated differently by the math layout engine and were inserted not as a `TextElem`, but as a `GlyphFragment` which created a frame based on its bounding box. Now a `TextElem` does the same thing and we don't have an inconsistency between spacing rules for one character and for a run of two or more.

As for `\alignlimits`, kinda cool, I didn't know about it.  To be fair, though, that's not a feature of vanilla LaTeX, but is a 2004 feature from `mathtools` which is an extension of `amsmath`, which sits on top of LaTeX (1984).  Typst still needs more work before something similar could be provided by a third party package.  A better version of `measure` is needed that returns baseline info about its argument, not just width and height.  But a better `measure` will come, too.

**Comment by laurmaedje** [comment]
What if it was just one option that does the right thing for top-edge and bottom-edge respectively. E.g. `#set text(top-edge: "bounds", bottom-edge: "bounds")`?

We would still need to bikeshed the name: Of the top of might head, I can think of `"bounds"`, `"tight"`, `"bbox"`, and `"bounding-box"`.

**Comment by damaxwell** [comment]
I was away from internet for a bit there. Yes, a larger change would certainly allow a single `bounds` (or however you want to spell it) option. It still doesn't address the weirdness that you can have a top that is `descender` and a bottom that is `x-height`. Regardless, I'll revise with this change.

**Comment by laurmaedje** [comment]
If you want, you can also change it such that top-edge only allows baseline-and-above metrics and bottom-edge baseline-and-below.

**Comment by damaxwell** [comment]
This is updated. `text(top-edge:)` accepts `"bounding-box"` or a vertical metric of `"baseline"` or above.  Similarly for `bottom-edge`.  I have no strong opinions on how to spell `"bounding-box"`.  It's wordy, but explicit.  I kinda like `"tight"` as an alternative.

File: crates/typst-library/src/math/ctx.rs [hunk 1]
```diff
@@ -204,9 +205,11 @@ impl<'a, 'b, 'v> MathContext<'a, 'b, 'v> {
...
             let text: EcoString = text.chars().map(|c| style.styled_char(c)).collect();
             let text = TextElem::packed(text)
-                .styled(TextElem::set_top_edge(TextEdge::Bbox(BboxVertical::BboxTop)))
-                .styled(TextElem::set_bottom_edge(TextEdge::Bbox(
-                    BboxVertical::BboxBottom,
+                .styled(TextElem::set_top_edge(TextTopEdge::Metric(
+                    TopEdgeMetric::BoundingBox,
+                )))
+                .styled(TextElem::set_bottom_edge(TextBottomEdge::Metric(
+                    BottomEdgeMetric::BoundingBox,
```
Comment (damaxwell) [inline crates/typst-library/src/math/ctx.rs]: I'm not sure what the right way is to make a new `TextElem` and set style properties.  I'm pretty sure that what I'm doing here is not the recommended route.  I had tried
```rust
        let text = TextElem::new(text)
            .spanned(span)
            .with_top_edge(TextTopEdge::Metric(TopEdgeMetric::BoundingBox))
            .with_bottom_edge(TextBottomEdge::Metric(BottomEdgeMetric::BoundingBox))
            .pack();
```
but the edge settings didn't have any effect.  Ideally, one wants to have a default of bounding box edges for text in math mode, but also allow these default settings to be overridden via the style mechanism. But I haven't wrapped my head around the architecture yet and could use a hand here.

Comment (laurmaedje) [reply crates/typst-library/src/math/ctx.rs]: You've probably hit a weird edges case of how text elements work (they do not typically have inherent fields which `.with_` creates). The way you've done it is okay.

**Comment by laurmaedje** [comment]
Seems like we both pushed here a bit interleaved. But the diff looks fine. Is this ready to merge?

**Comment by damaxwell** [comment]
There was a test image that was out of date after the recent merges to `main`.  I had rebased after your update and then fixed the image.  This all LGTM.

**Comment by laurmaedje** [comment]
Wonderful, thanks for working on this!

**Pull request merged**